### ISSUE TITLE
Improve AMPI error handling

### ISSUE THREAD

**Issue by stwhite91** [UIUC-PPL/charm#1095]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/1095*

---
Two related issues:
1. Error handling in AMPI is inconsistent: need to systematically review and add error codes to AMPI_* routines.
2. We currently use a mix of abort on failure (MPI_ERRORS_ARE_FATAL) and return on error (MPI_ERRORS_RETURN). The default should be to abort on error.

For #2, we must keep track of user-registered errhandler functions per comm/win/type, and invoke that function on error. The default one should print the name of the function that failed as well as the string returned by MPI_Error_string and abort. Error codes also belong to error classes, but I think we can just say that each error code is its own error class?

Basically, instead of returning an error code we should have something like 'return AMPI_ERRHANDLER("MPI_Send", comm, error_code);'

**Comment by stwhite91** [comment]
*Original date: 2016-09-08 12:55:29*

---
Added some rudimentary support for this here: ~~https://charm.cs.illinois.edu/gerrit/#/c/1828/~~ https://github.com/UIUC-PPL/charm/commit/8301f5fa41b9a407a2f900d83fbaa876cfd74910

**Comment by stwhite91** [comment]
*Original date: 2016-09-13 22:48:08*

---
Made AMPI abort on more types of errors and to be more informative when aborting: ~~https://charm.cs.illinois.edu/gerrit/#/c/1848/~~ https://github.com/UIUC-PPL/charm/commit/76d6c43e36ea8b16e44ef8f6f5ec3bb053c5677b

**Comment by stwhite91** [comment]
*Original date: 2017-01-12 05:21:41*

---
THings are better now for 6.8.0, but more systematic changes can be made for future releases.

**Comment by stwhite91** [comment]
*Original date: 2017-01-17 04:33:42*

---
The derived datatypes routines and RMA all need error handling.
